core/src/main/java/jeeves/transaction/TransactionManager.java

```diff
@@ -96,92 +112,309 @@
-                doRollback(context, transactionManager, transaction);
+                Log.error(Log.JEEVES, "Unexpected problem completing transaction, will try to rollback", t);
+                isRolledBack = rollbackIfNotRolledBack(context, transactionManager, status, isRolledBack);
+            }
+            Log.debug(
+                Log.JEEVES,
+                "Run in transaction completed:" +
+                    (isRolledBack?" rolled back":"") +
+                    (isCommitted?" committed":"")
+            );
+            if( isNewTransaction || commitBehavior == CommitBehavior.ALWAYS_COMMIT ) {
+                // this transaction was our responsibility, double checking with committed or rolledback successfully
+                if (!isRolledBack && !isCommitted) {
+                    Log.warning(Log.JEEVES, "Run in transaction did not complete cleanly, transaction not committed or rolledback");
+                }
+                if (isRolledBack && isCommitted) {
+                    Log.warning(Log.JEEVES, "Run in transaction did not complete cleanly, transaction both committed and rolledback");
+                }
             }
         }
 
-        if (exception[0] != null) {
-            if (exception[0] instanceof RuntimeException) {
-                throw (RuntimeException) exception[0];
-            } else if (exception[0] instanceof Error) {
-                throw (Error) exception[0];
-            } else {
-                throw new RuntimeException(exception[0]);
+    }
+
+    private static Boolean rollbackIfNotRolledBack(ApplicationContext context, PlatformTransactionManager transactionManager, TransactionStatus status, Boolean isRolledBack) {
+        if (!isRolledBack) {
+            doRollback(context, transactionManager, status);
+        }
+        return true;
+    }
+
+    /**
+     * Safely perform {@link PlatformTransactionManager#commit(TransactionStatus)}.
+     *
+     * This method is responsible for safely logging any notification failures as warnings, rather than failing outright.
+     *
+     * @param context context used to obtain listeners to notify
+     * @param transactionManager manager responsible for execution
+     * @param status transaction status
+     */
+    protected static void doCommit(ApplicationContext context, PlatformTransactionManager transactionManager, TransactionStatus status) {
+        try {
+            fireBeforeCommit(context, transactionManager, status);
+        }
+        catch( Throwable t ){
+            Log.warning(Log.JEEVES, "Commit transaction - before:", t);
+        }
+
+        if ( status == null || status.isCompleted()){
+            // not calling return here to preserve previous logic
+            // we can check if the following log messages are seen before taking defensive action
+            Log.debug( Log.JEEVES,"transactionManager.commit called unexpectedly when transaction is already completed ");
```

Comment:
I did not want to be agressive with calling out invalid state like this while doing improvements, since this is on master we could change this to a warning or even failure.